Rebase road property speeds

TheGreatRefrigerator opened the pull request:
### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box
when you have checked you have done them): -->
- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [ ] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [ ] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [ ] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [ ] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes # .

### Information about the changes
- Key functionality added:
- Reason for change:

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

Commit:
Reimplement Properties as enums

move to switch statements for readability

koebi committed:
feat: enhance RoadPropertySpeedParser

- add handling of unknown key
- throw StatusCodeException on validation error

TheGreatRefrigerator commented on openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java:
```diff
@@ -344,6 +347,18 @@ else if (ALT_ROUTE.equalsIgnoreCase(algoStr))
+						AbstractSpeedCalculatorWeighting weighting1 = (AbstractSpeedCalculatorWeighting) weighting;
+						if (weighting1.getSpeedCalculator() instanceof RoadPropertySpeedCalculator) {
+							((RoadPropertySpeedCalculator) weighting1.getSpeedCalculator()).setRoadPropertySpeedMap((RoadPropertySpeedMap) ((ORSPMap) request.getAdditionalHints()).getObj("user_speeds"));
+						}
```
i've also changed `weigthing1` to `weighting` here but without any difference.
Not sure which one is correct but i assume `weighting`. Probably was used to switch between the different weightings locally?

aoles commented:
After a quick look it seems to me that you are trying to re-implement the idea of using a speed calculator within the weightings. However, this concept has been established in the GH fork and in the ORS codebase already a while ago in https://github.com/GIScience/graphhopper/commit/68a5fa7eb0b5ca211687d98674b9bee324220189 and https://github.com/GIScience/openrouteservice/commit/ea68ed407adc876fbf2c8501f44ad6ea09d20046, https://github.com/GIScience/openrouteservice/commit/535d6ca8436a0c718babf91c5ae9caee3194a5b9, respectively. This change was primary implemented to deal with traffic-dependent speeds or time-dependent speed limits. 

Could it be that some of your code is just obsolete and is not integrating well with the existing functionality? Note that as currently any given weighting has an interface to hook-up a speed calculator, it should be enough just to define your new calculators and plug them in into any of the existing weightings such as with `MaximumSpeedCalculator` or `RoutingTrafficSpeedCalculator` without the need of any dedicated weightings. Depending on whether you would like to have a default fallback speed calculator in the back you might actually want to extend the `AbstractAdjustedSpeedCalculator`.

koebi commented:
Hey,
great finds!!
Turns out, "a while ago" (i.e. May/June 21) is indeed after the above changes have been implemented, which was before February 21.

Yeah, I'd assume that rewriting this to use the existing mechanism makes sense - however, I'm a bit confused since the code actually looks like it is doing that?

aoles committed:
Compare scaled speed values

Speed values stored on egdes are already scaled by a factor of 0.9. In order to reliably override them by user-provided values the latter ones need to be scaled as well before doing the comparison.

TheGreatRefrigerator committed:
feat: Add road property speed limits

- Add API Parameter 'user_speed_limits' to specify speed limits for road and surface types
- Add RoadPropertySpeedCalculator, ~Map and ~Parser
- Use custom speeds to adjust weighting in ORSGraphHopper.js
- Add RoadPropertySpeedMap to RouteSearchParameters

- Fix hasTimeDependentSpeedOrAccess variable casing

Co-authored-by: Hendrik Leuschner <hendrik@openrouteservice.org>
Co-authored-by: aoles <andrzej.oles@gmail.com>

TheGreatRefrigerator committed:
tests: Add RoadPropertyTest

Co-authored-by: Hendrik Leuschner <hendrik@openrouteservice.org>

TheGreatRefrigerator committed:
tests: Add user_speed_limits test

adds parameter and result tests for custom speeds

Co-authored-by: koebi <koebi@ezelo.de>

TheGreatRefrigerator commented:
I've force pushed(no diff) some squashing and restructuring of the commits so the review will be less of a torture.

But maybe @aoles and @koebi  you can take a quick look through the commits if it makes sense to you.

What's left to do is:
- some load testing which @koebi agreed to do (thanks).
- add changelog @TheGreatRefrigerator 
- check internal docs, comments and API playground @TheGreatRefrigerator

TheGreatRefrigerator committed:
feat: Add user_speed_limits parameter to isochrones endpoint

aoles commented:
@TheGreatRefrigerator Could you please remind me, what this PR was trying to address? 🧐

TheGreatRefrigerator commented:
> @TheGreatRefrigerator Could you please remind me, what this PR was trying to address? 🧐

Dynamic passing of speedlimits directly to the request it seems

aoles commented:
Thanks @TheGreatRefrigerator !

@MichaelsJP any thoughts how to deal with this? Unless we want to merge, I would propose to close the PR 💀

MichaelsJP commented:
Some explanation on this would have been good. @TheGreatRefrigerator is this still needed?

TheGreatRefrigerator commented:
> Some explanation on this would have been good. @TheGreatRefrigerator is this still needed?

Well, it would be one of the easiest ways for users to direclty adjust speeds, which makes adjustments for different countries a lot easier, than setting up ors and adjusting speed limits.

aoles commented:
@TheGreatRefrigerator Where does this come from though: was it an internal/external project, and does any documentation of it exist outside of this thread? 🤔

TheGreatRefrigerator commented:
Closed. If we want to integrate this again at some point, the branch is still there
